torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -118,14 +119,18 @@ c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
 
     futureResponseMessage->addCallback(
         [jitFuture](const FutureMessage& futureResponseMessage) {
-          // Don't need to acquire GIL here, as toPyObj acquires GIL
-          // when creating the py::object
           if (futureResponseMessage.hasError()) {
             jitFuture->setError(futureResponseMessage.error()->what());
           } else {
-            jitFuture->markCompleted(jit::toIValue(
-                toPyObj(futureResponseMessage.constValue()),
-                PyObjectType::get()));
+            IValue value;
+            {
+              // Need GIL to destruct the py::object returned by toPyObj
```

Comment:
I see.

Before https://github.com/pytorch/pytorch/pull/37519,
the code was like this, https://github.com/pytorch/pytorch/blame/580928801f5d4c229cb33b117f152af46e7a2a82/torch/csrc/distributed/rpc/init.cpp#L340
The `py::object` returned by `toPyObj(..)` was passed to Python directly. pybind11 takes care of its destruction.

While after the PR, this `toPyObj(..)` returned `py::object` is passed to `toIValue(..)` as `const py::handle&`, from which `PyObjectHolder` creates a new `py::object` copy. https://github.com/pytorch/pytorch/blob/dc28784f50da8276d87bfb7ed0b990aa88bc4ff0/torch/csrc/jit/python/python_ivalue.h#L20

With the destruction of the new copy taken care of by https://github.com/pytorch/pytorch/blob/dc28784f50da8276d87bfb7ed0b990aa88bc4ff0/torch/csrc/jit/python/python_ivalue.h#L29.

Now we have to take care of destruction of the old copy.

Reply:
Yes, exactly! 

There might be other places that need some change, e.g., UnpickledPythonCall holds a py::object, but relies on caller to call the `movePythonUdf`. Will fix this on top of this PR.